lld/ELF/SyntheticSections.h

```diff
@@ -788,6 +792,136 @@ class RelroPaddingSection final : public SyntheticSection {
   void writeTo(uint8_t *buf) override {}
 };
 
+class DebugNamesSection final : public SyntheticSection {
+  // N.B. Everything in this class assumes that we are using DWARF32.
+  // If we move to DWARF64, most of this data will need to be re-sized,
+  // and the code that handles or manipulates it will need to be updated
+  // accordingly.
+
+public:
+  DebugNamesSection();
+  template <typename ELFT> static DebugNamesSection *create();
+  void writeTo(uint8_t *buf) override;
+  size_t getSize() const override { return sectionSize; }
+  bool isNeeded() const override;
+
+  void addSections(SmallVector<InputSectionBase *, 0> sec_list) {
+    inputDebugNamesSections = sec_list;
+  }
+
+  template <class ELFT> void writeToImpl(uint8_t *buf);
+
+  template <class ELFT, class RelTy>
+  void getNameRelocsImpl(InputSection *sec, ArrayRef<RelTy> rels,
+                         llvm::DenseMap<uint32_t, uint32_t> &relocs);
+
+  template <class ELFT>
+  void getNameRelocs(InputSectionBase *base,
+                     llvm::DenseMap<uint32_t, uint32_t> &relocs);
+
+  template <class ELFT>
+  void endianWrite(uint8_t size, uint8_t *buf_start, uint32_t offset,
+                   uint32_t data);
+
+  struct Abbrev : public llvm::FoldingSetNode {
+    uint32_t code;
+    uint32_t tag;
+    SmallVector<llvm::DWARFDebugNames::AttributeEncoding, 2> attributes;
+
+    void Profile(llvm::FoldingSetNodeID &id) const;
+  };
+
+  struct AttrValueData {
+    uint32_t attrValue;
+    uint8_t attrSize;
+  };
+
+  struct IndexEntry {
+    uint32_t abbrevCode;
+    uint32_t poolOffset;
+    union {
+      int32_t parentOffset = -1;
```

Comment:
Can we use `uint32_t parentOffset = 0` since the minimum valid `parentOffset` value is at least `EntriesOffset`, which is much larger than 0?

Reply:
No, the offsets are measured from the start of the entries section, so zero is in fact a valid offset.  I need to initialize this to an invalid offset value.

Reply:
The header has a larger size, so offset 0 cannot really a valid entry offset.

Reply:
You would be correct, if the offsets were counting from the start of the .debug_names section, but they are not. The offsets are measuring from the start of the Entry Pool (the bit in .debug_names section that comes after the abbrev table).  So zero is a valid entry offset.

Reply:
Yep entry offsets are relative, with first one being zero for the purposes of parent. Discovered that when implementing support in BOLT.

Reply:
We compute the parent offset as `ie->parentOffset = locs.EntriesBase + attr.attrValue`. `EntriesBase` is larger than the header size. Here we can use 0 as a sentinel number.